src/gdextension/render.cpp

```diff
@@ -0,0 +1,389 @@
+
+        godot::Ref<godot::Image> img =
+            godot::Image::create_from_data(imgData.size().x(),
+                                           imgData.size().y(),
+                                           false,
+                                           formatMToGd(imgData.format()),
+                                           pba);
+        godot::RID rid = rs->texture_2d_create(img);
+        rRenderGd.m_texGd.emplace(newId, rid);
+    }
+}
+
+void SysRenderGd::compile_resource_meshes(
+    ACtxDrawingRes const &rCtxDrawRes, Resources &rResources, RenderGd &rRenderGd)
+{
+    // TODO: Eventually have dirty flags instead of checking every entry.
+
+    for ( [[maybe_unused]] auto const &[_, scnOwner] : rCtxDrawRes.m_meshToRes )
+    {
+        ResId const meshRes      = scnOwner.value();
+
+        // New element will be emplaced if it isn't present yet
+        auto const [it, success] = rRenderGd.m_resToMesh.try_emplace(meshRes);
+        if ( ! success )
+        {
+            continue;
+        }
+
+        // New element emplaced, this means we've just found a resource that
+        // isn't synchronized yet.
+
+        // Create new Mesh GL Id
+        MeshGdId const newId     = rRenderGd.m_meshIds.create();
+
+        // Create owner, this adds to the resource's reference count
+        ResIdOwner_t renderOwner = rResources.owner_create(restypes::gc_mesh, meshRes);
+
+        // Track with two-way map and store owner
+        rRenderGd.m_meshToRes.emplace(newId, std::move(renderOwner));
+        it->second                 = newId;
+
+        // Get mesh data
+        MeshData const &meshData   = rResources.data_get<MeshData>(restypes::gc_mesh, meshRes);
+        auto            primitive  = primitiveMToGd(meshData.primitive());
+
+        godot::RenderingServer *rs = godot::RenderingServer::get_singleton();
+
+        godot::SurfaceTool      st;
+        // why are there two different PrimitiveType enums ?
+        st.begin(static_cast<godot::Mesh::PrimitiveType>(primitive));
+
+        godot::RID mesh = rs->mesh_create();
+        auto pos = meshData.positions3DAsArray();
+        // TODO copy other attributes as well maybe
+        auto indices = meshData.indicesAsArray();
+        //TODO do that using an iterator I guess.
+        uint32_t sg = 0;
+        for ( auto i = indices.end(); i >= indices.begin() + 1;)
+        {
+            st.set_smooth_group(sg++);
+            auto v = pos[*(--i)];
+            st.add_vertex(godot::Vector3(v.x(), v.y(), v.z()));
+            v = pos[*(--i)];
+            st.add_vertex(godot::Vector3(v.x(), v.y(), v.z()));
+            v = pos[*(--i)];
+            st.add_vertex(godot::Vector3(v.x(), v.y(), v.z()));
+        }
+        if ( primitive == godot::RenderingServer::PRIMITIVE_TRIANGLES )
+        {
+            //st.deindex();
+            st.generate_normals();
+            //st.generate_tangents();
+            
+        }
+
+        godot::Array meshArray = st.commit_to_arrays();
```

Comment:
This is the SurfaceTool that's leaking. Calling `clear` does not change anything.

Reply:
I presume a RID is being leaked? though I notice there's no mention of RIDs in `godot/scene/resources/surface_tool.cpp`

Reply:
Unrelated to the leak: the way the indexed mesh is being turned into just vertices is throwing off the smooth shading. It looks like there's methods for directly adding an index buffer to SurfaceTool. I implied this somewhat in a previous comment, but it's likely that SurfaceTool isn't needed entirely.

Reply:
It's not a RID, it's the object itself. There is apparently an "ObjectDB" in Godot, and it detects that all the surface tool used to create meshes are still allocated when the app closes. 

I'll try tinkering it a bit, but I remember getting weird results when using other ways to create meshes.

Reply:
> Unrelated to the leak: the way the indexed mesh is being turned into just vertices is throwing off the smooth shading. It looks like there's methods for directly adding an index buffer to SurfaceTool. I implied this somewhat in a previous comment, but it's likely that SurfaceTool isn't needed entirely.

It turns out not ignoring normals fixes the smooth shading. Who could have thought.